File: kotlin/src/main/kotlin/200.NumberofIslands.kt
import org.junit.Assert.assertEquals
import org.junit.Test
import java.util.*
import kotlin.collections.ArrayList

//https://leetcode.com/problems/number-of-islands/

class NumberofIsland{
  fun visit(grid: Array<CharArray>, i: Int, j: Int){
    val que = Stack<Pair<Int, Int>>()
    que.push(i to j)
    while(que.isNotEmpty()){
      val (i,j) = que.pop()
      grid[i][j] = 'x'
      for((di, dj) in listOf(0 to 1, 1 to 0, 0 to -1, -1 to 0)){
        val ni = i + di
        val nj = j + dj
        if(0<=ni && ni<grid.size && 0<=nj && nj<grid[i].size && grid[ni][nj] == '1')
          que.push(ni to nj)
      }
    }
  }

  fun numIslands(grid: Array<CharArray>): Int {
    var count = 0
    for(i in 0 until grid.size)
      for(j in 0 until grid[i].size)
        if(grid[i][j] == '1'){
          count ++
          visit(grid, i, j)
        }
    return count
  }

  @Test
  fun test1(){
    assertEquals(1, numIslands(arrayOf(
        "11110",
        "11010",
        "11000",
        "00000"
    ).map{it.toCharArray()}.toTypedArray()))
  }

  @Test
  fun test2(){
    assertEquals(3, numIslands(arrayOf(
        "11000",
        "11000",
        "00100",
        "00011"
    ).map{it.toCharArray()}.toTypedArray()))
  }
}

File: python/200_NumberofIslands.py
from unittest import TestCase
# https://leetcode.com/problems/number-of-islands/


class NumberofIslands(TestCase):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] == "1":
                    count += 1
                    self.visit(i, j, grid)
        return count

    def visit(self, i, j, grid):
        que = [(i, j)]
        while len(que)>0:
            i, j = que.pop()
            grid[i][j] = -1
            for (di, dj) in ((0, 1), (1, 0), (0, -1), (-1, 0)):
                ni = i+di
                nj = j+dj
                if 0 <= ni < len(grid) and 0 <= nj < len(grid[ni]) and grid[ni][nj] == "1":
                    que.append((ni, nj))

    # # union-find
    # def numIslands(self, grid):
    #     root, cnt = {}, 0
    #
    #     def find(pos):
    #         if pos not in root: return None
    #         if root[pos] != pos: root[pos] = find(root[pos])
    #         return root[pos]
    #     for i in range(len(grid)):
    #         for j in range(len(grid[i])):
    #             if grid[i][j] == "0": continue
    #             cnt += 1
    #             pos = (i, j)
    #             root[pos] = pos
    #             for nei in (j-1 >= 0 and (i, j-1), i-1 >= 0 and (i-1, j)):
    #                 if not nei: continue
    #                 rn, rp = find(nei), find(pos)
    #                 if rn and rn != rp:
    #                     root[rp] = rn
    #                     cnt -= 1
    #     return cnt

    def test1(self):
        grid = [
            list("11110"),
            list("11010"),
            list("11000"),
            list("00000"),
        ]
        self.assertEqual(1, self.numIslands(grid))

    def test2(self):
        grid = [
            list("11000"),
            list("11000"),
            list("00100"),
            list("00011"),
        ]
        self.assertEqual(3, self.numIslands(grid))


